go-controller/pkg/ovn/port_cache.go

```diff
@@ -0,0 +1,93 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/klog"
+)
+
+type portCache struct {
+	sync.RWMutex
+
+	counter  uint64
+	stopChan <-chan struct{}
+	cache    map[string]*lpInfo
+}
+
+type lpInfo struct {
+	name          string
+	uuid          string
+	logicalSwitch string
+	ip            net.IP
+	mac           net.HardwareAddr
+	// dead indicates that this object is scheduled to be removed
+	dead uint64
+}
+
+func newPortCache(stopChan <-chan struct{}) *portCache {
+	return &portCache{
+		stopChan: stopChan,
+		cache:    make(map[string]*lpInfo),
+	}
+}
+
+func (c *portCache) get(logicalPort string) (*lpInfo, error) {
+	c.RLock()
+	defer c.RUnlock()
+	if info, ok := c.cache[logicalPort]; ok {
+		klog.V(5).Infof("port-cache(%s): port found in cache", logicalPort)
+		return info, nil
+	}
+	return nil, fmt.Errorf("logical port %s not found in cache", logicalPort)
+}
+
+func (c *portCache) add(logicalSwitch, logicalPort, uuid string, mac net.HardwareAddr, ip net.IP) *lpInfo {
+	c.Lock()
+	defer c.Unlock()
+	portInfo := &lpInfo{
+		logicalSwitch: logicalSwitch,
+		name:          logicalPort,
+		uuid:          uuid,
+		ip:            ip,
+		mac:           mac,
+	}
+	klog.V(5).Infof("port-cache(%s): added port %+v", logicalPort, portInfo)
+	c.cache[logicalPort] = portInfo
+	return portInfo
+}
+
+func (c *portCache) remove(logicalPort string) {
+	c.Lock()
+	defer c.Unlock()
+	info, ok := c.cache[logicalPort]
+	if !ok || info.dead > 0 {
+		return
+	}
+	info.dead = atomic.AddUint64(&c.counter, 1)
```

Comment:
I think it may be a better idea here to use a real timestamp rather than a counter. Since you are doing cache expiry here based on time, we can use a real time stamp which gets updated. Then you can just compare that in your goroutine, and set it to nil or something when the port is re-added. I think it makes the logic easier and debugging will be easier. Similar to how mac-learning tables work in L2 switches.

Reply:
@trozet can do that, and I pushed a fixup for your consideration. But it's basically the same as the stamp except more complicated because now we have to compare times rather than just a number.

Reply:
I originally had a single goroutine that periodically searched the map and cleaned up anything that had expired (using time.Time expired value in each port object compared against time.Now()) but girish pointed out that might not scale well with thousands of ports.